enterprise/coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,511 @@
+	case <-ctx.Done():
+		// nolint:gocritic // it's okay to use slog.F() for an error in this case
+		// because we want to differentiate two different types of errors: ctx.Err() and context.Cause()
+		c.logger.Error(context.Background(), "reconciler stop exited prematurely", slog.Error(ctx.Err()), slog.F("cause", context.Cause(ctx)))
+	// Wait for the control loop to exit.
+	case <-c.done:
+		c.logger.Info(context.Background(), "reconciler stopped")
+	}
+}
+
+func (c *StoreReconciler) isStopped() bool {
+	return c.stopped.Load()
+}
+
+// ReconcileAll will attempt to resolve the desired vs actual state of all templates which have presets with prebuilds configured.
+//
+// NOTE:
+//
+// This function will kick of n provisioner jobs, based on the calculated state modifications.
+//
+// These provisioning jobs are fire-and-forget. We DO NOT wait for the prebuilt workspaces to complete their
+// provisioning. As a consequence, it's possible that another reconciliation run will occur, which will mean that
+// multiple preset versions could be reconciling at once. This may mean some temporary over-provisioning, but the
+// reconciliation loop will bring these resources back into their desired numbers in an EVENTUALLY-consistent way.
+//
+// For example: we could decide to provision 1 new instance in this reconciliation.
+// While that workspace is being provisioned, another template version is created which means this same preset will
+// be reconciled again, leading to another workspace being provisioned. Two workspace builds will be occurring
+// simultaneously for the same preset, but once both jobs have completed the reconciliation loop will notice the
+// extraneous instance and delete it.
+func (c *StoreReconciler) ReconcileAll(ctx context.Context) error {
+	logger := c.logger.With(slog.F("reconcile_context", "all"))
+
+	select {
+	case <-ctx.Done():
+		logger.Warn(context.Background(), "reconcile exiting prematurely; context done", slog.Error(ctx.Err()))
+		return nil
+	default:
+	}
+
+	logger.Debug(ctx, "starting reconciliation")
+
+	err := c.WithReconciliationLock(ctx, logger, func(ctx context.Context, db database.Store) error {
+		snapshot, err := c.SnapshotState(ctx, db)
+		if err != nil {
+			return xerrors.Errorf("determine current snapshot: %w", err)
+		}
+		if len(snapshot.Presets) == 0 {
+			logger.Debug(ctx, "no templates found with prebuilds configured")
+			return nil
+		}
+
+		var eg errgroup.Group
+		// Reconcile presets in parallel. Each preset in its own goroutine.
+		for _, preset := range snapshot.Presets {
+			ps, err := snapshot.FilterByPreset(preset.ID)
+			if err != nil {
+				logger.Warn(ctx, "failed to find preset snapshot", slog.Error(err), slog.F("preset_id", preset.ID.String()))
+				continue
+			}
+
+			if !preset.UsingActiveVersion && len(ps.Running) == 0 && len(ps.InProgress) == 0 {
+				logger.Debug(ctx, "skipping reconciliation for preset; inactive, no running prebuilds, and no in-progress operations",
```

Comment:
This short-circuit is not necessary: ReconcilePreset should accurately calculate the actions.

I don't think it's a good idea to have this short circuit because you're splitting the business logic of reconciliation into two different places that do the same thing today. If we want to modify the logic, it will be easy to make the change only in one place and have inconsistent or undesired behavior.

Reply:
Okay, I'll remove it
cc @dannykopping